packages/foam-vscode/src/core/services/markdown-link.ts

```diff
@@ -15,6 +15,7 @@ export abstract class MarkdownLink {
           link.rawText
         );
         return {
+          rawTarget: target,
```

Comment:
have you considered changing the regex to capture the alias separator used and returning that?
I would say that semantically might be a slightly "cleaner" solution, as the `\` is not really part of the target.

What do you think?

Reply:
I had tried that, but I found there has a so uncommon use case is `\` in [valid  character in file name on linux / Unix(MacOS) but not in windows.](https://en.wikipedia.org/wiki/Filename) 

But currently, the oddly character link `|` `\` will break down the of Foam filename analyze system a lot. If there are not going to dealing with these cases. I prefer the cleaner solution you said that use regex to split the alias separator and combined it.

Reply:
Not sure if I fully understand, let me give you some context.
In Foam identifiers only use the `/` divider to describe a path (maybe `\` currently works but that's a coincidence). e.g. when we create a unique identifier, which needs to include part of the path, we use `/`.
Also, the only thing we care about is if the `\` char is used right before the `|`. It's a bit of an odd case (because in theory it only applies inside tables), but I think it's ok to generalize to the fact that the allowed separators for the alias are `|` and `\|`. 
Does the above make sense? Is that answering your questions?

Reply:
I got the point. but there is another problem.

I found the case: If use `\` in filename (this is so weird, and is invalid in windows). In this case, we cannot realize the `\` char is the part of escaped in divider or filename, only if extract the special case regex of escaped divider inside table, but it will be more complicated (need another special regex rule I think)

But if does not care this case(its uncommon and only occur in Linux and Unix).  we can just improve the current regex like what you said.

Reply:
If I understand, your special case only applies when `\` is the last char of the filename.
In that case, but really even in the case of `\` anywhere in the file, I feel this is an uncommon enough exception that I am ok with not supporting it.

I was thinking, there might be one more issue with having `\` in the target: will these two links point to the same file?

---

| col1 | col2 |
|-|-|
| [[wikilink\|alias]] | [[wikilink]] |

[[wikilink]]

---

I have a feeling the link with the escape would resolve to another target?
It would actually be good to have a test case for this just to be sure.
But yeah, that feels like the risk of "overloading" the `target` and the why it feels "safer" to simply expose the `aliasSeparator`.

Thoughts?

Reply:
> I was thinking, there might be one more issue with having \ in the target: will these two links point to the same file?

Do you mean in the source text is

```
| col1            | col2     |
| --------------- | -------- |
| [[wikilink\|alias]] | [[wikilink]] |

[[wikilink]]
```

In the cases above,  it will link to the same file `wikilink.md`. But consider the cases below. 

```
file name with '|'
[[test|]]                                  // link to test.md

file name with '\'
[[test\]]                                  // link to test.md

alias link file name with '\'
[[test\|test]]                             // link to test.md

section link file name with '\'
[[test\#Graph Visualization]]              // link to test.md

section alias link file name with '\'
[[test\#Graph Visualization|test]]         // link to test.md

|                                      | 
| -----------------------------------  | 
| [[test\]]                            |   // link to test.md
| [[test\\|alias]]                     |   // cannot be parse as valid wikilink
| [[test\#Graph Visualization\|test]]  |   // link to test.md
| [test](test\.md)                     |   // link to test\.md
| [test](test\.md#custom-graph-styles) |   // link to test\.md
| [test](test|.md)                     |   // link to test|.md
| [test](test|.md#custom-graph-styles) |   // link to test|.md
```

These case is not complete. but the uncommon end char will break the link parser. I'm agree with `expose the aliasSeparator` is  the better way. 

So I think we should have two kind of `aliasSeperator`? 

- `\|` only use in table for escape. 
- `|` use in any other context

And in regex maybe we should split out these case.

Reply:
Thanks for the recap, it helped me think about this holistically, here are my thoughts: the way I see it is just that we need to support `\|`, for a very specific case, which is an alias inside a table.
I don't feel we need to be strict about where the link is (that would get messy pretty quickly), so what I would say in general is that `\|` is equivalent to `|`.
To me that's all we need, which means that your table above to me should be: 

```
file name with '|'
[[test|]]                                  // link to test.md

file name with '\'
[[test\]]                                  // link to test\.md

alias link file name with '\'
[[test\|test]]                             // link to test.md

section link file name with '\'
[[test\#Graph Visualization]]              // link to test\.md

section alias link file name with '\'
[[test\#Graph Visualization|test]]         // link to test\.md

|                                      | 
| -----------------------------------  | 
| [[test\]]                            |   // link to test\.md
| [[test\\|alias]]                     |   // link to test\.md
| [[test\#Graph Visualization\|test]]  |   // link to test\.md
| [test](test\.md)                     |   // link to test\.md
| [test](test\.md#custom-graph-styles) |   // link to test\.md
| [test](test|.md)                     |   // link to test|.md
| [test](test|.md#custom-graph-styles) |   // link to test|.md
```

I know it's slightly different from the current implementation, unsure whether this simplify things or not, but I do believe it's the "most appropriate" interpretation of the issue at hand, which to rephrase it for me is about supporting wikilink alias inside a table, and not supporting escape chars for section/alias separators. What do you think?

Reply:
>  I would say they are both valid and when refactoring we keep track of which one was used and keep using it. Some might say this is not 100% correct, and they might be right, but it's simple for the user to understand and I'd say it covers 99.998% of cases.

I'm agree with the simplicity, and its is enough that cover most of use cases

> supporting wikilink alias inside a table, and not supporting escape chars for section/alias separators. 

Is this mean we should deal with incorrect case in current implementation below: 

```
[[test\\|alias]]                       |   // cannot be parse as valid wikilink => link to test\.md
| [[test\#Graph Visualization\|test]]  |   // link to test.md => link to test\.md
```

In addition, I would like to ask where I can start to tracking the wikilnck is better than the beginning, It will make me resolve the problem more efficiently

Reply:
> Is this mean we should deal with incorrect case in current implementation below:

What do you mean by "incorrect"?
In those cases I would say

`[[test\\|alias]]` maps to `test\` (given that here the separator is `\|`)
`[[test\#graph|test]]` maps to `test\` because we only support `\|` and no other "special use" of `\`

is that answering the question?

> In addition, I would like to ask where I can start to tracking the wikilnck is better than the beginning, It will make me resolve the problem more efficiently

Not sure I understand, what do you mean?

Thanks!

Reply:
I'm sorry for unclear description. 

In this PR, I only fix the issue that update wikilink after rename file. 
But on the discussion above, we should also fix the problem that parser parse the wikilink in table with escape char in these case. 

```
| [[test\\|alias]]                     |   
| [[test\#Graph Visualization\|test]]  |
```

I'm willing to deal with it. But if any info like mentioned in [this comment ](https://github.com/foambubble/foam/issues/990#issuecomment-1112350100) will be helpful.

Reply:
IIUC the approach of changing `MarkdownLink.analyzeLink` to support the two alias separators (`\|` and `|`) and report which one is used should fix both getting the right identifier within a table and the file rename. If I am missing something I would say let's focus this PR on the issue of updating wikilink on file rename (as you said), and if the approach feels incompatible with the second objective we can review it, I hope this makes sense.